p4runtime_sh/test.py

```diff
@@ -1109,6 +1110,48 @@ def test_global_options_invalid(self):
         with self.assertRaisesRegex(UserError, "Invalid value type"):
             sh.global_options["canonical_bytestrings"] = "bar"
 
+    def test_packet_in(self):
+        # In this tests we will send a packet-in message from the servicer and check if
+        # packet_in.sniff method works
+        msg = p4runtime_pb2.StreamMessageResponse()
+        msg.packet.payload = b'Random packet-in payload'
+        md = p4runtime_pb2.PacketMetadata()
+        md.metadata_id = 1
+        md.value = b'\x00\x01'
+        msg.packet.metadata.append(md)
+
+        # Have to sniff the packet in another thread since this blocks the thread
+        packet_in = sh.PacketIn()
+        captured_packet = []
+
+        def _sniff_packet(captured_packet):
+            captured_packet += packet_in.sniff(timeout=1)
+        _t = Thread(target=_sniff_packet, args=(captured_packet, ))
+        _t.start()
+
+        # TODO: modify the servicer to send stream message?
+        sh.client.stream_in_q["packet"].put(msg)
+        _t.join()
+
+        self.assertEqual(len(captured_packet), 1)
+        self.assertEqual(captured_packet[0],  msg)
+
+    def test_packet_out(self):
+        expected_msg = p4runtime_pb2.StreamMessageRequest()
+        expected_msg.packet.payload = b'Random packet-out payload'
+        md = p4runtime_pb2.PacketMetadata()
+        md.metadata_id = 1
+        md.value = b'\x00\x01'
+        expected_msg.packet.metadata.append(md)
+
+        packet_out = sh.PacketOut()
+        packet_out.payload = b'Random packet-out payload'
+        packet_out.metadata['egress_port'] = '1'
+        packet_out.send()
+
+        actual_msg = sh.client.stream_out_q.get()
```

Comment:
how can this work consistently? I would expect the thread executing `stream_req_iterator` in `P4RuntimeClient` to potentially remove the item from the queue (and send it to the server) before we can retrieve it here. Isn't there a potential race condition?

Reply:
Right, I think we should store the packet-out message in the mock server(Servicer) and get the message from there.